spec/lib/puppet_spec/https.rb

```diff
@@ -0,0 +1,166 @@
+require 'spec_helper'
+require 'webrick'
+
+class PuppetSpec::HTTPSServer
+  attr_reader :ca_cert, :ca_crl, :server_cert, :server_key
+
+  def initialize
+    @ca_cert = OpenSSL::X509::Certificate.new(CA_CERT)
+    @ca_crl = OpenSSL::X509::CRL.new(CRL)
+    @server_key = OpenSSL::PKey::RSA.new(SERVER_KEY)
+    @server_cert = OpenSSL::X509::Certificate.new(SERVER_CERT)
+    @config = WEBrick::Config::HTTP.dup
+  end
+
+  def handle_request(ctx, ssl)
+    req = WEBrick::HTTPRequest.new(@config)
+    req.parse(ssl)
+
+    res = WEBrick::HTTPResponse.new(@config)
+    res.status = 200
+    res.body = 'OK'
+    res.send_response(ssl)
+  end
+
+  def start_server(&block)
+    errors = []
+
+    IO.pipe {|stop_pipe_r, stop_pipe_w|
+      store = OpenSSL::X509::Store.new
+      store.add_cert(@ca_cert)
+      store.purpose = OpenSSL::X509::PURPOSE_SSL_CLIENT
+      ctx = OpenSSL::SSL::SSLContext.new
+      ctx.cert_store = store
+      ctx.cert = @server_cert
+      ctx.key = @server_key
+      ctx.verify_mode = OpenSSL::SSL::VERIFY_NONE
+
+      Socket.do_not_reverse_lookup = true
+      tcps = TCPServer.new("127.0.0.1", 0)
+      begin
+        port = tcps.connect_address.ip_port
+        begin
+          server_thread = Thread.new do
+            begin
+              ssls = OpenSSL::SSL::SSLServer.new(tcps, ctx)
+              ssls.start_immediately = true
+
+              loop do
+                readable, = IO.select([ssls, stop_pipe_r])
+                break if readable.include? stop_pipe_r
+
+                ssl = ssls.accept
+                begin
+                  handle_request(ctx, ssl)
+                ensure
+                  ssl.close
+                end
+              end
+            rescue => e
+              # uncomment this line if something goes wrong
+              # puts "SERVER #{e.message}"
+              errors << e
+            end
+          end
+
+          begin
+            yield port
+          ensure
+            stop_pipe_w.close
+          end
+        ensure
+          server_thread.join
+        end
+      ensure
+        tcps.close
+      end
+    }
+
+    errors
+  end
+
+CA_CERT = <<END
```

Comment:
Is it reasonable to pass these artifacts in on initialize so that they can come from fixtures?  They need to match the ones used by the client, right?

Reply:
> They need to match the ones used by the client, right?

It's more the other way around. The server can use whatever certs/keys it wants, but in order for the client to authenticate the server, then the client needs to trust the server's CA cert "out of band". But the CA cert is the only shared state between the two.

The client may not have a client cert, or it can use the server's cert, which simulates an agent running on the puppetserver's host.

Reply:
So should both this server, and the client being tested, use CA files from [here](https://github.com/puppetlabs/puppet/blob/3d046400a87ff1afe10fdd511e3530ff19766d5e/tasks/generate_cert_fixtures.rake#L48)?